lib/ansible/modules/remote_management/manageiq/manageiq_tenant.py

```diff
@@ -0,0 +1,447 @@
+#!/usr/bin/python
+#
+# (c) 2018, Evert Mulder (base on manageiq_user.py by Daniel Korn <korndaniel1@gmail.com>)
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import (absolute_import, division, print_function)
+
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+
+module: manageiq_tenant
+
+short_description: Management of tenants in ManageIQ.
+extends_documentation_fragment: manageiq
+version_added: '2.7'
+author: Evert Mulder
+description:
+  - The manageiq_tenant module supports adding, updating and deleting tenants in ManageIQ.
+
+options:
+  state:
+    description:
+    - absent - tenant should not exist, present - tenant should be.
+    choices: ['absent', 'present']
+    default: 'present'
+  name:
+    description:
+      - The tenant name.
+    required: true
+    default: null
+  description:
+    description:
+    - The tenant description.
+    required: true
+    default: null
+  parent_id:
+    description:
+    - The id of the parent tenant. If not supplied the root tenant is used.
+    required: false
+    default: null
+  quotas:
+    description:
+    - The tenant quotas.
+    - All parameters case sensitive.
+    - 'Valid attributes are:'
+    - ' - C(cpu_allocated) (int): use null to remove the quota.'
+    - ' - C(mem_allocated) (GB): use null to remove the quota.'
+    - ' - C(storage_allocated) (GB): use null to remove the quota.'
+    - ' - C(vms_allocated) (int): use null to remove the quota.'
+    - ' - C(templates_allocated) (int): use null to remove the quota.'
+    required: false
+    default: null
+'''
+
+EXAMPLES = '''
+- name: Update the root tenant in ManageIQ
+  manageiq_tenant:
+    name: 'My Company'
+    description: 'My company name'
+    manageiq_connection:
+      url: 'http://127.0.0.1:3000'
+      username: 'admin'
+      password: 'smartvm'
+      verify_ssl: False
+
+- name: Create a tenant in ManageIQ
+  manageiq_tenant:
+    name: 'Dep1'
+    description: 'Manufacturing department'
+    parent_id: 1
```

Comment:
Is 1 assumed to be the `My Company` tenant?  how can one know it's 1, without prior assumptions and by only using this module?
=> Can you make the module take `parent_name` param and look it up, instead of id?

Reply:
I like the path idea.  `/root/sub1/sub2` is much easier than calling the module 3 times, with `register:` to get parent ids to follow the chain...
Is `/` (or another char) impossible in a name?  If not, could take a list of components (and maybe allow a string `'foo'` as shorthand for `['foo']` for convenience).

Great question about renaming!  I don't think we supported renaming in any other miq module.   @gtanzillo @abellotti @zgalor @yaacov @elad661 any ideas?
Perhaps add an `old_name` or `new_name` (or path) field.  Is there any existing pattern for this in Ansible?  Can it be idempotent?

Reply:
I know in the vmware_guest lets you supply the VM uuid to rename a VM, see https://docs.ansible.com/ansible/2.5/modules/vmware_guest_module.html, search for "Rename a VM (requires the VM's uuid)". So perhaps to support renaming the internal Id could be supplied.

If have tested if a special character can be used as a separator, but all chars slashes, backslashes are all valid for a tenant name. Having looked at the vmware_guest module they use the name_match attribute, in combination with the uuid. This could also be an approach:
1. First use tenant_id, if supplied (rename tenant if name differs)
2. Use parent_id, if supplied , and tenant name 
3. Use tenant name, If multiple found use the name_match attribute to select the first found or last found

or use the tenant_path as a list. I don't think the shorthand if 1 name is supplied would work. It is unclear if you want to edit the root tenant or some other sub tenant i think

Reply:
I ended up allowing the giving the option to supply the parent name with build-in uniqueness checks as an alternative to the parent_id. This should cover most situations I think.